Ctr drbg variable key size support

NirSonnenschein opened the pull request:
Notes:
* Pull requests cannot be accepted until:
-  The submitter has [accepted the online agreement here with a click through](https://developer.mbed.org/contributor_agreement/) 
   or for companies or those that do not wish to create an mbed account, a slightly different agreement can be found [here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/)
- The PR follows the [mbed TLS coding standards](https://tls.mbed.org/kb/development/mbedtls-coding-standards)
* This is just a template, so feel free to use/remove the unnecessary things
## Description
allow different sized keys in the CTR DRBG API. this is configured at compile time.
256 bit keys are still default as before. Warning will be displayed if 128bit keys are chosen.


## Status
**IN DEVELOPMENT**

## Requires Backporting
When there is a bug fix, it should be backported to all maintained and supported branches.
Changes do not have to be backported if:
- This PR is a new feature\enhancement
- This PR contains changes in the API. If this is true, and there is a need for the fix to be backported, the fix should be handled differently in the legacy branch

NO  
Which branch?

## Migrations
If there is any API change, what's the incentive and logic for it.

YES 
There is an implicit change in an API (CTR-DRBG will now allow use with 128bit keys [configuration at compile time]). motivation is allowing the use of some (hardware) implementations which are 128bit only.

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported


## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

NirSonnenschein commented:
Hi @Patater, 
here is the Pull request for the different sized key support in CTR DRBG.
Please take a look when you have a chance, any inputs on what could be wrong with the new test vector will be very helpful.
Note: the test vector for 128bit keys are still not passing (hence the DO NOT MERGE label).

RonEld commented:
@NirSonnenschein Please fix CI failures.

Where di you get the test vectors for the AES128 key size? It seems to be identical to another test vector for AES256 key size. See [here](https://github.com/ARMmbed/mbedtls/pull/1902/files#diff-146c97196e1e1245a509f6358dd3c2baR3).
It is unreasonable that same test data will work for both key sizes

NirSonnenschein commented:
Thanks @RonEld , the test vector was from the NIST website you showed me.
I'll re-check this again. Looking into the CI issues now.

NirSonnenschein commented:
retest

NirSonnenschein commented:
Added additional test Vectors for AES-128 based CTR-DRBG. 
Thanks to @gilles-peskine-arm who contributed the test modifications making this possible.

NirSonnenschein commented:
retest

NirSonnenschein commented:
Thanks again to @gilles-peskine-arm  for the contribution of the test changes.

Patater commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -8,8 +8,11 @@
+ * as the underlying block cipher.
+ * 
+ *  * \warning   ARC4 is considered a weak cipher and its use constitutes a
+ *            security risk. We recommend considering stronger ciphers instead.
```
Style: funny misaligned wrapping here

Patater commented on ChangeLog:
```diff
@@ -49,6 +49,8 @@ Features
    * Add support for key wrapping modes based on AES as defined by
      NIST SP 800-38F algorithms KW and KWP and by RFC 3394 and RFC 5649.
+   * Add support different keys sizes to CTR_CRBG. Note: using shorter keys
+     than 256 bit may decrease security.
```
Grammar: use "bits" instead of "bit".

It'd be more natural to say "using keys shorter than" instead of "using short keys than".

Patater commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -8,8 +8,11 @@
- * block cipher.
+ * The Mbed TLS implementation of CTR_DRBG uses AES-256 (default) or AES-128 
+ * as the underlying block cipher.
+ * 
```
Don't add trailing whitespace in commit "enable the use of variable size keys in ctr_drbg (was hard-coded 256bit before)".

Patater commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -52,7 +52,7 @@
 #else
 #if defined(MBEDTLS_CTR_DRBG_KEY_SIZE_128)
-#warning Warning: using smaller (128bit) key size for CTR DRBG may reduce the security of some operations.
+#pragma message("Warning: using smaller (128bit) key size for CTR DRBG may reduce the security of some operations.")
```
Does this message pragma work on all our supported toolchains? Is there a reason we can't use the same mechanism that `MBEDTLS_TEST_NULL_ENTROPY` uses for its warning?

Patater commented on library/ctr_drbg.c:
```diff
@@ -66,6 +66,18 @@ void mbedtls_ctr_drbg_init( mbedtls_ctr_drbg_context *ctx )
  * Non-public function wrapped by mbedtls_ctr_drbg_seed(). Necessary to allow
```
Commit "ctr_drbg: add comments relating the code with the NIST specification" doesn't (and other commits that were cherry-picked don't) need to say "cherry-picked from ...", especially not from a commit that was never part of a public long-lived branch. The author information is still showing @gilles-peskine-arm as author, so why mention the cherry pick?

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.data:
```diff
@@ -1,723 +1,1075 @@
+CTR_DRBG_withDF.pdf: AES-256, PR=no, perso=no, add=no
+depends_on:MBEDTLS_CTR_DRBG_KEY_SIZE_256
```
You need to update `MBEDTLS_CTR_DRBG_KEY_SIZE_256` to `!MBEDTLS_CTR_DRBG_USE_128_BIT_KEY` and `MBEDTLS_CTR_DRBG_KEY_SIZE_128` to `MBEDTLS_CTR_DRBG_USE_128_BIT_KEY`.

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.data:
```diff
@@ -1,723 +1,1075 @@
 
 CTR_DRBG NIST Validation (AES-256 use df,False,256,128,0,0) #7
+depends_on:MBEDTLS_CTR_DRBG_KEY_SIZE_256
 ctr_drbg_validate_nopr:"a2d5eff6f73f98e5b04c01967dffa69b":"59759bb91b3c4feb18c0f08626
```
What is this lowercase letter after the closing parenthesis?

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -2910,6 +2910,7 @@
 //#define MBEDTLS_CTR_DRBG_MAX_INPUT                256 /**< Maximum number of additional input bytes */
 //#define MBEDTLS_CTR_DRBG_MAX_REQUEST             1024 /**< Maximum number of requested bytes per call */
 //#define MBEDTLS_CTR_DRBG_MAX_SEED_INPUT           384 /**< Maximum size of (re)seed buffer */
+//#define MBEDTLS_CTR_DRBG_USE_128_BIT_KEY              /** use 128 bit key for CTR-DRBG - redcuces security */
```
The comment needs to explain in what way this reduces security. If the explanation is long, this can be a reference to another file such as `ctr_drbg.h`.

Patater commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -54,83 +63,107 @@ exit:
+                                  strlen( entropy_string ) / 2, entropy_string,
+                                  "", add1_string, add2_string,
+                                  result_string );
+    goto exit;
```
Why do we need these "goto exit;" in all the tests?

gilles-peskine-arm replied:
The root cause is that the function `ctr_drbg_validate` is both used as a test function, and as an auxiliary function with some trivial wrappers. That's not something the test framework supports well. It's a test function so it returns `void`. The trivial wrappers are left with no source of failure, so no natural `goto exit` (which is normally hidden inside a `TEST_ASSERT`). But the test framework adds an `exit` label to test functions if they don't have one already, and many compilers complain about an unused label.

I think it would be best to change `ctr_drbg_validate` into an auxiliary function that returns `int`, and to have four simple wrapper functions (one for each `RESEED_xxx`) that each call `TEST_ASSERT( ctr_drbg_validate( … ) )`. This means creating a wrapper function for `RESEED_FIRST`. I didn't have time to do this on Friday.

Patater commented on include/mbedtls/config.h:
```diff
@@ -2920,6 +2910,7 @@
 //#define MBEDTLS_CTR_DRBG_MAX_INPUT                256 /**< Maximum number of additional input bytes */
 //#define MBEDTLS_CTR_DRBG_MAX_REQUEST             1024 /**< Maximum number of requested bytes per call */
 //#define MBEDTLS_CTR_DRBG_MAX_SEED_INPUT           384 /**< Maximum size of (re)seed buffer */
+//#define MBEDTLS_CTR_DRBG_USE_128_BIT_KEY              /** use 128 bit key for CTR-DRBG - redcuces security */
```
Do we need to update all the test vector dependencies now as well (to use MBEDTLS_CTR_DRBG_USE_128_BIT_KEY instead of MBEDTLS_CTR_DRBG_KEY_SIZE_128)?

NirSonnenschein replied:
yes, fixed.

NirSonnenschein commented on ChangeLog:
```diff
@@ -49,6 +49,8 @@ Features
      independently contributed again by Paul Sokolovsky.
    * Add support for key wrapping modes based on AES as defined by
      NIST SP 800-38F algorithms KW and KWP and by RFC 3394 and RFC 5649.
+   * Add support different keys sizes to CTR_CRBG. Note: using keys
```
fixed

NirSonnenschein commented on include/mbedtls/config.h:
```diff
@@ -2018,7 +2018,9 @@
  * \def MBEDTLS_CTR_DRBG_C
  *
- * Enable the CTR_DRBG AES-256-based random generator.
+ * Enable the CTR_DRBG AES-based random generator.
```
fixed

NirSonnenschein replied:
changed all instances back to CTR_DRBG

NirSonnenschein commented on include/mbedtls/config.h:
```diff
@@ -2910,6 +2912,7 @@
 //#define MBEDTLS_CTR_DRBG_MAX_INPUT                256 /**< Maximum number of additional input bytes */
 //#define MBEDTLS_CTR_DRBG_MAX_REQUEST             1024 /**< Maximum number of requested bytes per call */
 //#define MBEDTLS_CTR_DRBG_MAX_SEED_INPUT           384 /**< Maximum size of (re)seed buffer */
+//#define MBEDTLS_CTR_DRBG_USE_128_BIT_KEY              /** use 128 bit key for CTR-DRBG - may reduce security (see ctr_drbg.h) */
```
fixed

NirSonnenschein commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -45,7 +48,13 @@
 
 #define MBEDTLS_CTR_DRBG_BLOCKSIZE          16 /**< The block size used by the cipher. */
+
+#if !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* use default 256-bit key for CTR-DRBG.*/
```
fixed

NirSonnenschein commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -8,8 +8,11 @@
+ * as the underlying block cipher.
+ *
+ *  * \warning values generated by AES-128 based CTR_DRBG may be less secure
+ *  * as they are easier to guess.
```
fixed

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -2018,14 +2018,16 @@
  * Requires: MBEDTLS_AES_C
  *
- * This module provides the CTR_DRBG AES-256 random number generator.
+ * This module provides the CTR-DRBG AES-256 random number generator.
```
The standard spelling is `CTR_DRBG`. `CTR-DRBG` is a common spelling which is used in some Mbed TLS documentation, but NIST spells it with an underscore.

gilles-peskine-arm commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -45,7 +48,13 @@
+#if !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* Use default 256-bit key for CTR-DRBG. */
 #define MBEDTLS_CTR_DRBG_KEYSIZE            32 /**< The key size used by the cipher. */
+#else
+#pragma message("Warning: using smaller (128-bit) key size for CTR DRBG may reduce the security of some operations.")
```
At a glance it seems that `#pragma message` is supported by reasonably modern versions of GCC, Clang, Visual and IAR. Those are the compilers that we absolutely must support. Have you checked whether it's supported by the specific versions we need to support?

I'd still prefer to use the more common `#warning`, and wrap it with `#if defined(MBEDTLS_DEPRECATED_WARNING)`. Since we recommend against using this feature, it is functionally deprecated.

NirSonnenschein replied:
Hi @gilles-peskine-arm, 
I did a little digging in regards to #pragma #warning
1. I found that GCC versions equal or greater than 4.4.7 (version 4.3.6 doesn't) of GCC support #pragma message, (https://gcc.gnu.org/onlinedocs/gcc-4.4.7/gcc/Diagnostic-Pragmas.html#Diagnostic-Pragmas)
2. ancedotal search evidence indicates this is supported in visual studio 2010.
3. for clang it should be part of version 2.9 (https://github.com/llvm-mirror/clang/blob/release_29/test/Lexer/pragma-message.c)
4. for IAR version 5.4 supports it (couldn't find docs for earlier versions)(http://supp.iar.com/FilesPublic/UPDINFO/004916/arm/doc/EWARM_DevelopmentGuide.ENU.pdf)

If this level of support is not sufficient I can change the messages back to #warning (with the downsides noted above).

gilles-peskine-arm commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -45,7 +48,13 @@
+#if !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* Use default 256-bit key for CTR-DRBG. */
 #define MBEDTLS_CTR_DRBG_KEYSIZE            32 /**< The key size used by the cipher. */
+#else
+#pragma message("Warning: using smaller (128-bit) key size for CTR DRBG may reduce the security of some operations.")
```
Since we warn against using 128-bit CTR_DRBG, why are we not officially marking it as `\deprecated`?

NirSonnenschein replied:
I'm not sure what is the policy regarding this feature is. Normally when I have seen a feature referred to as sepracated this means that it will soon be removed and users should switch to using something else instead. This might be confusing to put on a new feature/configuration option.

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -2018,14 +2018,16 @@
  * \def MBEDTLS_CTR_DRBG_C
  *
- * Enable the CTR_DRBG AES-256-based random generator.
+ * Enable the CTR-DRBG AES-based random generator.
```
It's spelled `CTR_DRBG`, not `CTR-DRBG`. You don't need to fix pre-existing occurrences, but please don't introduce new occurrences of `CTR-DRBG`.

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.data:
```diff
@@ -1,723 +1,1075 @@
 
 CTR_DRBG NIST Validation (AES-256 use df,False,256,128,0,0) #5
+depends_on:!MBEDTLS_CTR_DRBG_USE_128_BIT_KEY
 ctr_drbg_validate_nopr:"ae0b0d2e84f48c632f031356cdea60ac":"554cf6fad1c376ad6148cd40b53105c16e2f5dd5fa564865b26faa8c318150bfb2294e711735df5eb86ff4b4e778531793bad42403d93a80d05c5421229a53da":"":"":"":"1212e5d3070b1cdf52c0217866481c58"
```
I still don't understand why there's a bunch of test cases with an extra lowercase letter after the closing parenthesis. Please explain it or remove it.

NirSonnenschein replied:
The explanation is simple, The letter is just a differentiation between the tests. There are multiple test vectors with the same input configuration, so the test title would be exactly the same if not for the letter suffix. this would make it difficult to figure out which test failed in case of a test failure.
If you think a different marker would be clearer, I'd be happy to change this to something else (I didn't want to use a number as this might get confused with the iteration number #X at the end of the test name).

gilles-peskine-arm replied:
The `#X` numbers already serve this purpose, and unlike the letter it's clear what they mean. It's better to keep the number aligned. How about `CTR_DRBG CAVS 14.3 (…) block #1 COUNT=0`?

NirSonnenschein replied:
I though the #X numbers reffered to somthing else (speicifcally the COUNT = X value in the test data). 
In the CAVS there existis the following format:
1. test parameters definitions (AES type, df, pr, ) e.g. 
[AES-256 use df]
[PredictionResistance = False]
[EntropyInputLen = 256]
[NonceLen = 128]
[PersonalizationStringLen = 0]
[AdditionalInputLen = 0]
[ReturnedBitsLen = 512]

2. then a set of test data blocks enumerated by COUNT = X (I though this X corresponds to the #X in the test name)

I noticed that there are in fact more than one instance of test parameter definitions with the same definitions. for example, the parameters below appear 4 times (hence a, b, c, and d), each with it's own test data labeled count of COUNT=0 to COUNT= 14:
[AES-256 use df]
[PredictionResistance = False]
[EntropyInputLen = 256]
[NonceLen = 128]
[PersonalizationStringLen = 0]
[AdditionalInputLen = 0]
[ReturnedBitsLen = 512]

I tired to include a value from each of these blocks, but I had to differentiate between them (all the test parameters are the same). using the #X which corresponded to count seemed to be misleading since the tests may have the same COUNT value.

NirSonnenschein replied:
I've adopeted something like your suggestion:
CTR_DRBG CAVS 14.3 (AES-256 no df,no reseed,256,128,0,0) block 1 #0

that way the # will mean the same thing as in the previous tests, with just the block as additional information.

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -50,76 +63,126 @@ exit:
+                                  strlen( entropy_string ) / 4, entropy_string,
+                                  add_reseed_string, add1_string, add2_string,
+                                  result_string );
+    goto exit;
```
Not a blocker, but I'd prefer to change `ctr_drbg_validate` into an auxiliary function that returns `int`, and to have four simple wrapper functions (one for each `RESEED_xxx`) that each call `TEST_ASSERT( ctr_drbg_validate( … ) )`. See https://github.com/ARMmbed/mbedtls/pull/1902#discussion_r207929721

gilles-peskine-arm commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -8,8 +8,11 @@
+ * as the underlying block cipher.
+ *
+ *  \warning values generated by AES-128 based CTR_DRBG may be less secure
+ *           as they are easier to guess.
```
This warning is not informative. I think we want something like this — @yanesca please comment and improve.

> Using AES-128 for the random number generator limits the security of all generated keys and all operations that use random values to 128-bit security.

Do we need to invoke the birthday paradox as well?

NirSonnenschein replied:
I'm more than happy to adopt your version. I'll wait to see inputs from @yanesca  before updating to avoid multiple updates

yanesca replied:
I am happy with the wording. Birthday attacks don't apply here, but even if they would, the reseed limit would prevent them (our reseed limit is even more conservative than the one prescribed by NIST).

NirSonnenschein replied:
Thanks @yanesca  and @gilles-peskine-arm , changed to text suggested by Gilles.

gilles-peskine-arm commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -45,7 +48,13 @@
+#if !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* Use default 256-bit key for CTR-DRBG. */
 #define MBEDTLS_CTR_DRBG_KEYSIZE            32 /**< The key size used by the cipher. */
+#else
+#pragma message("Warning: using smaller (128-bit) key size for CTR DRBG may reduce the security of some operations.")
```
Please clarify “may reduce the security of some operations”. I'd say something like “limits the strength of generated material to 128 bits”. See the discussion about the wording of `\warning` above.

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -49,6 +49,8 @@ Features
      independently contributed again by Paul Sokolovsky.
    * Add support for key wrapping modes based on AES as defined by
      NIST SP 800-38F algorithms KW and KWP and by RFC 3394 and RFC 5649.
+   * Add support for different keys sizes to CTR_DRBG. Note: using keys
```
Not “different key sizes”, just 128. Something like (@yanesca @Patater):

> Add support for 128-bit keys in CTR_DRBG. Note that using keys shorter than 256 bits limits the security of generated material to 128 bits.

NirSonnenschein replied:
I'm more than happy to use your phrasing. I'll wait to see if there are additional comments from @Patater or @yanesca before updating to avoid multiple updates.

yanesca replied:
I am happy with the wording here too.

NirSonnenschein replied:
Thanks @yanesca  and @gilles-peskine-arm , changed to text suggested by Gilles.

NirSonnenschein commented:
Hi @Patater, by the MBEDTLS_TEST_NULL_ENTROPY comment do you mean:
1.  adding a warning directly in the Makefile (Line 57-70) / CMakeLists.txt (Lines 27-50)?
2.  issuing a warning if MBEDTLS_TEST_NULL_ENTROPY is defined (entropy.c lines 30-34)
I can use one or both, depending on your feedback but they do come with a few drawbacks.

option 2, is simple but has the two drawbacks:
1. possible issues with CI and warnings may prevent tesitn 128-bit keys in CI
2. customers will not be able to compile code with warnings as errors and use 128-bit keys.

options 1 doesn't have these drawbacks but:
1. is a bit more complicated 
2. is inapplicable if the customer doesn't use the provided build system.

also @gilles-peskine-arm and @Patater  any other issues with this PR.

Patater commented:
Hi @NirSonnenschein 

I don't see a strong reason to be different than our existing MBEDTLS_TEST_NULL_ENTROPY mechanism, as it was designed with very similar constraints and requirements in mind. The pragma is likely to get lost and ignored, and might not be so portable. A warning prevents building for a valid target with warnings as errors.

To address your concern about when users don't use our build system, for Mbed OS at least, our importer script modifies the `config-no-entropy.h` configuration file to do a `#warning`. We could take a similar approach with this 128-bit option. This isn't any worse off than the `MBEDTLS_TEST_NULL_ENTROPY` warning system we already have.

NirSonnenschein commented:
Thanks @Patater, 
based on your answer I understand that the preferred approach is warning in the Makefile / CMake.
I'm making the modifications for this mechanism now.

NirSonnenschein commented:
Hi @Patater @gilles-peskine-arm , 
I've changed the warning mechanism to use the "makefile/Cmake print" print instead of the pragma.
please let me know if there are any further issues.
Thanks

Patater commented on CMakeLists.txt:
```diff
@@ -36,9 +36,26 @@ set(NULL_ENTROPY_WARNING "${WARNING_BORDER}"
 find_package(Perl)
 if(PERL_FOUND)
 
+    # If NULL Entropy is configured, display an appropriate warning
```
Copy-paste error: this comment should say "If CTR_DRBG is configured to use 128-bit keys" or something like that.

Patater commented on Makefile:
```diff
@@ -61,9 +61,21 @@ NULL_ENTROPY_WARN_L3=****  AND IS *NOT* SUITABLE FOR PRODUCTION USE     ****\n
+WARNING_BORDER_LONG      =****************************************************************************\n
+CTR_DRBG_128_BIT_KEY_WARN_L1=****  WARNING!  MBEDTLS_CTR_DRBG_USE_128_BIT_KEY defined!               ****\n
+CTR_DRBG_128_BIT_KEY_WARN_L2=****  Using 128-bit keys for CTR_DRBG limits the security of generated  ****\n
+CTR_DRBG_128_BIT_KEY_WARN_L3=****  keysand operations that use random values to 128-bit security     ****\n
```
"keysand" -> "keys and"
"random values"->"generated material" or "generated values" (as @yanesca prefers)

Is there a reason this text is different from the CMake version? I like the Makefile version better, but please use whatever @yanesca recommends for this text as the wording needs to be very particular. All three warnings: the doxygen "\warning", the CMake, and the Makefile should probably all be the same exact wording.

@yanesca Please recommend the exact text we should use for the warnings.

Patater commented on Makefile:
```diff
@@ -61,9 +61,21 @@ NULL_ENTROPY_WARN_L3=****  AND IS *NOT* SUITABLE FOR PRODUCTION USE     ****\n
 post_build:
 ifndef WINDOWS
+
+	# If 128-bit keys are configured for CTR DRBG, display an appropriate warning
```
"CTR DRBG" -> CTR_DRBG

Patater commented on CMakeLists.txt:
```diff
@@ -36,9 +36,26 @@ set(NULL_ENTROPY_WARNING "${WARNING_BORDER}"
                          "${WARNING_BORDER}")
 
+set(CTR_DRBG_128_BIT_KEY_WARN_L1 "****  WARNING!  MBEDTLS_CTR_DRBG_USE_128_BIT_KEY defined!\n")
+set(CTR_DRBG_128_BIT_KEY_WARN_L2 "****  Using 128-bit keys for CTR DRBG limits the security of all generated keys\n")
```
"CTR DRBG" -> "CTR_DRBG"

yanesca commented:
All three changes to the original text make sense and do not change the meaning.

NirSonnenschein commented:
fixed warnings to all be the same (commit not registering in GitHub PR yet).

gilles-peskine-arm commented:
This will need rebasing to the tip of `development` since some aspects of test functions have changed. When you do this, please also remove the function `ctr_drbg_validate` and instead use `ctr_drbg_validate_no_reseed` for `RESEED_NEVER` and a new wrapper function `ctr_drbg_validate_reseed_first` for `RESEED_FIRST`.

NirSonnenschein commented:
Hi @Patater and @gilles-peskine-arm, 
I've rebased this on the current development branch and solved the naming issues created.
The branch prior to rebase can be found here for comparison:
https://github.com/NirSonnenschein/mbedtls/tree/CTR-DRBG_variable_key_size_support_before_rebase .
please take a look when you have a chance.

Patater commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -187,8 +187,8 @@ void ctr_drbg_validate_pr( char *add_init_string, char *entropy_string,
 
 /* BEGIN_CASE */
-void ctr_drbg_validate_nopr( data_t * add_init, data_t * entropy,
-                             data_t * add1, data_t * add_reseed,
```
Why isn't data_t available to use?

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -51,50 +144,57 @@ exit:
-                             data_t * add2, data_t * result_str )
+void ctr_drbg_validate_no_reseed( char *add_init_string, char *entropy_string,
+                                  char *add1_string, char *add2_string,
+                                  char *result_string )
```
The input data arguments are byte arrays, not null-terminated strings. They must have the type `data_t *` and you must use `foo->len`, not `strlen`.

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -51,50 +144,57 @@ exit:
-void ctr_drbg_validate_pr( data_t * add_init, data_t * entropy,
-                           data_t * add1, data_t * add2,
-                           data_t * result_str )
+void ctr_drbg_validate( int reseed_mode, char *nonce_string,
```
Please move this function to the header section, create a fourth wrapper function for `RESEED_FIRST`, and call the wrapper functions from the `.data` file instead of calling `ctr_drbg_validate` directly.

NirSonnenschein commented:
Hi @gilles-peskine-arm @Patater, 
I've fixed the tests so that they use data_t as parameters and rebased to clean up the commit history as requested (previous branch can be found here for comparison). @gilles-peskine-arm I'm not sure what you mean by the missing wrapper function: the code (even prior to the current rebase) uses a single common utility function and 4 wrapper functions for the different scenarios (there is no function used as both a test and a utility function as before.). Am I missing something?

please take a look when you can.

itayzafrir commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -1,15 +1,96 @@
+                        int entropy_len_arg, data_t * entropy,
+                        data_t * reseed,
+                        data_t * add1, data_t * add2,
+                        data_t * result )
```
Style: pointer arguments should be declared as `data_t *result` instead of `data_t * result` (here and elsewhere).

itayzafrir commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -1,15 +1,96 @@
+};
+
+static size_t test_offset_idx;
+static size_t test_max_idx;
```
How about setting default values for `test_offset_idx` & `test_max_idx` so that the `if` will fail deterministically in case this function is called without setting values for these 2?

gilles-peskine-arm replied:
I wrote the code to figure out how to map the official test data to our function calls. I didn't write it for robustness. Any robustness or readability improvement is appreciated.

NirSonnenschein replied:
zeroed variables to make the test code more robust.

itayzafrir commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -1,15 +1,96 @@
+
+exit:
+    mbedtls_ctr_drbg_free( &ctx );
+    return 0;
```
Always `0`? Then what's the point of a return value?

gilles-peskine-arm replied:
A redundant goto statement is better than a function that always returns 0. At least the redundant goto statement is localized to the place where it's used, whereas making `ctr_drbg_update_internal` return a value also impacts `ctr_drbg_update_internal`.

NirSonnenschein replied:
based on @itayzafrir and @gilles-peskine-arm  comments, I've gone pack to the previous approach (with redundant goto, and added a comment explaning this)

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -23,80 +23,29 @@ static int mbedtls_test_entropy_func( void *data, unsigned char *buf, size_t len
-                        char *reseed_string,
-                        char *add1_string, char *add2_string,
-                        char *result_string )
+static int ctr_drbg_validate_internal( int reseed_mode, data_t * nonce,
```
Since this function calls `TEST_ASSERT` to report test failures, make it return `void`. Calling `TEST_ASSERT` is what marks the test as failing, you don't have to report it to the main function of the unit test if the main function doesn't need the information. I.e. go back to what it was before 6275be34831cb6b07c1bf326cd795750dbe8886f.

NirSonnenschein replied:
Fixed

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -109,15 +58,15 @@ void ctr_drbg_validate( int reseed_mode, char *nonce_string,
     }
 
-    /* CTR_DRBG_Generate(result_len * 8 bits, add1[:add1_len]) -> buf */
+    /* CTR_DRBG_Generate(result->len * 8 bits, add1[:add1_len]) -> buf */
```
Please change `add1_len` to `add1->len` as well. Same thing with the other `xxx_len`.

NirSonnenschein replied:
fixed

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -23,80 +23,29 @@ static int mbedtls_test_entropy_func( void *data, unsigned char *buf, size_t len
     unsigned char buf[64];
-    unsigned char result[64];
+
     size_t entropy_chunk_len = (size_t) entropy_len_arg;
```
Please add `TEST_ASSERT( entropy_chunk_len <= sizeof( buf ) )` so we don't get a nasty surprise later if we add a test with a different output size.

NirSonnenschein replied:
check added

gilles-peskine-arm commented on tests/suites/test_suite_ctr_drbg.function:
```diff
@@ -126,58 +75,111 @@ void ctr_drbg_validate( int reseed_mode, char *nonce_string,
+/* END_CASE */
+
+/* BEGIN_CASE */
+void ctr_drbg_validate( int reseed_mode, data_t * nonce,
```
Please remove this function. In the test data, replace `ctr_drbg_validate:RESEED_NEVER` by `ctr_drbg_validate_no_reseed`, `ctr_drbg_validate:RESEED_ALWAYS` by `ctr_drbg_validate_pr`, and `ctr_drbg_validate:RESEED_FIRST` by `ctr_drbg_validate_reseed_first` (some subsequent arguments may need to be adjusted). Write `ctr_drbg_validate_reseed_first` on the same model as the other three. Rename `ctr_drbg_validate_nopr` to `ctr_drbg_validate_reseed_between` for consistency.

NirSonnenschein replied:
I've made the changes requested (thanks for the detailed explanation)

gilles-peskine-arm commented on library/ctr_drbg.c:
```diff
@@ -256,6 +268,14 @@ static int block_cipher_df( unsigned char *output,
 }
 
+/* CTR_DRBG_Update (SP 800-90A &sect;10.2.1.2)
+ * ctr_drbg_update_internal(ctx, provided_data)
```
`provided_data` is the name used in the standard. So are `CTR_DRBG_Update`, `Key` and `V`. The point of these comments is to relate the code to the terminology in the standard.

I think the comment as is is clear enough. If you want to align it with the parameter name then the comment should be
```
/* CTR_DRBG_Update (SP 800-90A &sect;10.2.1.2)
 * ctr_drbg_update_internal(ctx, data)
 * implements
 * CTR_DRBG_Update(provided_data, Key, V)
 * with inputs and outputs
 *   ctx->aes_ctx = Key
 *   ctx->counter = V
 *   data = provided_data
 */
```
but I think the additional indirection only makes it harder to read.

Patater commented on CMakeLists.txt:
```diff
@@ -36,10 +36,26 @@ set(NULL_ENTROPY_WARNING "${WARNING_BORDER}"
                          "${NULL_ENTROPY_WARN_L3}"
                          "${WARNING_BORDER}")
 
-find_package(PythonInterp)
```
Is it intentional to remove `find_package(PythonInterp)`?

NirSonnenschein replied:
no,  fixed.

NirSonnenschein commented:
Hi @gilles-peskine-arm and @Patater,
You probably won't have time today, but just in case I've fixed the issues raised by the latest reviews.

gilles-peskine-arm commented on include/mbedtls/config.h:
```diff
@@ -2018,7 +2018,9 @@
- * Enable the CTR_DRBG AES-256-based random generator.
+ * Enable the CTR_DRBG AES-based random generator.
+ * The CTR_DRBG generator uses AES-256 by default.
+ * To use AES-128 instead, enable MBEDTLS_CTR_DRBG_USE_128_BIT_KEY below.
```
In “This module provides the CTR_DRBG AES-256 random number generator.” below, please change “AES-256” to “AES”.

NirSonnenschein replied:
fixed

gilles-peskine-arm commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -45,7 +48,12 @@
 
 #define MBEDTLS_CTR_DRBG_BLOCKSIZE          16 /**< The block size used by the cipher. */
+
+#if !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* Use default 256-bit key for CTR_DRBG. */
```
Please avoid “if not condition then x else y” unless there's a good reason. Change to `#if defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY)` and swap the key sizes.

NirSonnenschein replied:
fixed

gilles-peskine-arm commented:
I have manually verified that the configuration default - `MBEDTLS_AESNI_C` + `MBEDTLS_CTR_DRBG_USE_128_BIT_KEY` passes the test suite with ASan, with MSan and with Valgrind.

gilles-peskine-arm commented on include/mbedtls/ctr_drbg.h:
```diff
@@ -49,10 +49,10 @@
-#if !defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* Use default 256-bit key for CTR_DRBG. */
-#define MBEDTLS_CTR_DRBG_KEYSIZE            32 /**< The key size used by the cipher. */
-#else
+#if defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY) /* Use default 256-bit key for CTR_DRBG. */
```
“Use default 256-bit key for CTR_DRBG” doesn't correspond to the case `defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY)`. It would be better if the typeset documentation reflected the choice of key size anyway. Suggestion:
```
#if defined(MBEDTLS_CTR_DRBG_USE_128_BIT_KEY)
#define MBEDTLS_CTR_DRBG_KEYSIZE            16 /**< The key size used by the cipher (compile-time choice: 128 bits). */
￼ #else
￼ #define MBEDTLS_CTR_DRBG_KEYSIZE            32 /**< The key size used by the cipher (compile-time choice: 256 bits). */
￼ #endif
```

NirSonnenschein replied:
good point, your version looks better. I'm amending the commit to reduce noise.

gilles-peskine-arm commented:
Approved by 2 reviewers, CI passed, and functionality tested manually. This is ready for merge.

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -71,6 +71,8 @@ Features
    * Add support for key wrapping modes based on AES as defined by
      NIST SP 800-38F algorithms KW and KWP and by RFC 3394 and RFC 5649.
+   * Add support for 128-bit keys in CTR_DRBG. Note that using keys shorter
+     than 256 bits limits the security of generated material to 128 bits.
```
Please add: “Fixes #1146.”, with a commit message containing “fix #1146” so that Github will close the issue when this PR is merge.
